library/L2_HAL/sensors/battery/ltc4150.hpp

```diff
@@ -4,95 +4,74 @@
 class Ltc4150 : public CoulombCounter
 {
  public:
-  /// struct for holding the number of times an ISR invoked for a coulomb
-  /// counter
-  struct Battery_t
+  /// Defines what state the Polarity pin can be in.
+  enum class Polarity : uint8_t
   {
-    /// int variable holding the number of times an ISR invoked for a coulomb
-    /// counter
-    int num_ticks;
+    kDischarging,
+    kCharging,
   };
-
-  /// Class that handles the ticks from the LTC4150. The class uses a template
-  /// so there can be different handlers for different LTC4150s connected to a
-  /// board.
-  template <int id>
-  class TickHandler
-  {
-   public:
-    /// This member variable will hold the number of ticks sent from the
-    /// LTC4150. The interrupt handler increments this value by one every time
-    /// it is invoked.
-    inline static Battery_t battery_info;
-
-    /// handles the ISR by incrementing the num_ticks variable of TickHandler's
-    /// Battery_t
-    static void IsrHandler()
-    {
-      battery_info.num_ticks += 1;
-    }
-
-    /// @return the calculated milliamp hours based on the number of ticks.
-    /// 3600 is X 32.55 is kGvF and 0.00005 is kRsense
-    static float GetBatterymAh()
-    {
-      return static_cast<float>(static_cast<float>(battery_info.num_ticks) /
-                                (3600 * kGvf * kRsense));
-    }
-  };
-
-  /// @param isr - handler for coulomb counter ticks.
   /// @param int_pin - gpio pin for handling handling interrupts from the
-  /// LTC4150
+  ///        LTC4150
   /// @param pol - gpio pin to determine the polarity output of the LTC4150.
-  template <int id>
-  explicit constexpr Ltc4150(const TickHandler<id> & isr,
-                             Gpio & int_pin,
-                             Gpio & pol)
+  /// @param resistance - value of impedance represented in ohms for
+  ///        calculating battery charge
+  explicit constexpr Ltc4150(sjsu::Gpio & int_pin,
+                             sjsu::Gpio & pol,
+                             units::impedance::ohm_t resistance)
       : int_pin_(int_pin),
-        pol_(pol),
-        get_mah_(isr.GetBatterymAh),
-        isr_handler_(isr.IsrHandler)
+        pol_pin_(pol),
+        resistance_(resistance),
+        pulses_(0),
+        polarity_(Polarity::kDischarging)
   {
   }
 
   /// Initialize hardware, setting pins as inputs and attaching ISR handlers to
   /// the interrupt pin.
-  void Initialize() const override
+  void Initialize() override
   {
-    pol_.SetAsInput();
+    pol_pin_.SetAsInput();
     int_pin_.SetAsInput();
-    int_pin_.AttachInterrupt(isr_handler_, Gpio::Edge::kEdgeFalling);
+    pol_pin_.AttachInterrupt(
+        [this]() {
+          polarity_ =
+              (pol_pin_.Read()) ? Polarity::kDischarging : Polarity::kCharging;
+        },
+        Gpio::Edge::kEdgeFalling);
+    int_pin_.AttachInterrupt([this]() { pulses_++; }, Gpio::Edge::kEdgeFalling);
```

Comment:
```suggestion
    int_pin_.AttachInterrupt([this]() { 
    if (polarity == Polarity::kDischarging)
    {
      pulses_--; 
     } 
     else 
     {
        pulses_++;
     }
    }, Gpio::Edge::kEdgeFalling);
```
The polarity can change at any time during usage of a coulomb counter thus, you need to check it to decide if you should decrement the pulses or increment.